sycl/include/sycl/ext/intel/experimental/esimd/kernel_properties.hpp

```diff
@@ -0,0 +1,78 @@
+//==---------------- common.hpp - DPC++ Explicit SIMD API   ----------------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+// APIs for setting kernel properties interpreted by GPU software stack.
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include <sycl/ext/intel/experimental/esimd/common.hpp>
+#include <sycl/ext/intel/experimental/esimd/detail/misc_intrin.hpp>
+
+#include <sycl/detail/boost/mp11.hpp>
+
+__SYCL_INLINE_NAMESPACE(cl) {
+namespace __ESIMD_ENS {
+
+namespace kernel_properties {
+
+/// <summary>
+///  This namespace contains APIs to set kernel properties.
+/// </summary>
+
+// Implementation note: <property_class>::value fields should match property IDs
+// specified in llvm/lib/SYCLLowerIR/ESIMD/LowerESIMD.cpp
+
+namespace detail {
+// Proxy to access private property classes' fields from the API code.
+template <class T> struct proxy {
+  static constexpr int value = T::value;
+};
+} // namespace detail
+
+/// A boolean property which requests the compiler to double the amount of
+/// general-purpose registers available to a thread at the expense of reducing
+/// the amount of available hardware threads.
+struct use_double_grf_tag {
+  template <class T> friend struct detail::proxy;
+
+private:
+  // Property identifier
+  static constexpr int value = 0;
+};
+
+inline constexpr use_double_grf_tag use_double_grf = {};
+
+} // namespace kernel_properties
+
+namespace __MP11_NS = sycl::detail::boost::mp11;
+
+// TODO this should be replaced with the generic SYCL compile-time properites
+// mechanism once implementaion is available.
+// https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/proposed/sycl_ext_oneapi_kernel_properties.asciidoc
+
+template <class... KernelProps>
+void set_kernel_properties(KernelProps... props) {
```

Comment:
Since there is no extension document for this API, I assume it is an unsupported API that we can remove without any deprecation period.  Is that your intent?

What is the `__ESIMD_ENS` namespace?  Will users know that APIs in this namespace are unsupported?  Should we instead define the API in a namespace named `experimental`?

Reply:
> Since there is no extension document for this API, I assume it is an unsupported API that we can remove without any deprecation period. Is that your intent?

We usually allow some deprecation period for ESIMD APIs being removed for users to adapt, even though it is experimental.

> What is the __ESIMD_ENS namespace?

this is an alias for sycl::ext::intel::experimental::esimd

> Will users know that APIs in this namespace are unsupported?

This is a good question. The main source of information about particular APIs for users should be the API docs doxygen https://intel.github.io/llvm-docs/doxygen/group__sycl__esimd.html. But it turns out it does not expand aliases, I we need to to replace aliases with normal nested namespace declaration style. Thanks for bringing this up.

Reply:
@gmlueck, this is fixed in https://github.com/intel/llvm/pull/6197 (doxygen has an option to expand)